core/blockchain_rollback_test.go

```diff
@@ -0,0 +1,432 @@
+// the side chain is abo e the commited block. In this case we expect the canonical
+// chain to be rolled back to the committed block, with everything afterwads deleted;
+// the side chain completely nuked by the freezer.
+func TestLongNewerForkedFastSyncedRepair(t *testing.T) {
+	testRepair(t, 24, 12, 16, 4, 4, 4, 0)
+}
+
+// Tests a recovery for a long canonical chain with frozen blocks and a shorter
+// side chain, where the fast sync pivot point - older than the ancient limit -
+// was not yet comitted, but the process crashed. In this test scenario the side
+// chain is abo e the commited block. In this case we expect the chain to detect
+// that it was fast syncing and not delete anything. The side chain is completely
+// nuked by the freezer.
+func TestLongNewerForkedFastSyncingRepair(t *testing.T) {
+	testRepair(t, 24, 12, 16, 0, 4, 24, 0)
+}
+
+// Tests a recovery for a long canonical chain with frozen blocks and a longer side
+// chain, where a recent block - older than the ancient limit - was already comitted
+// to disk and then the process crashed. In this case we expect the canonical chains
+// to be rolled back to the committed block, with everything afterwads deleted. The
+// side chain completely nuked by the freezer.
+//
+// The side chain could be left to be if the fork point was before the new head
+// we are deleting to, but it would be exceedingly hard to detect that case and
+// properly handle it, so we'll trade extra work in exchange for simpler code.
+func TestLongReorgedRepair(t *testing.T) {
+	testRepair(t, 24, 26, 16, 4, 0, 4, 0)
+}
+
+// Tests a recovery for a long canonical chain with frozen blocks and a longer
+// side chain, where the fast sync pivot point - older than the ancient limit -
+// was already comitted to disk and then the process crashed. In this case we
+// expect the canonical chains to be rolled back to the committed block, with
+// everything afterwads deleted. The side chain completely nuked by the freezer.
+//
+// The side chain could be left to be if the fork point was before the new head
+// we are deleting to, but it would be exceedingly hard to detect that case and
+// properly handle it, so we'll trade extra work in exchange for simpler code.
+func TestLongReorgedFastSyncedRepair(t *testing.T) {
+	testRepair(t, 24, 26, 16, 4, 4, 4, 0)
+}
+
+// Tests a recovery for a long canonical chain with frozen blocks and a longer
+// side chain, where the fast sync pivot point - older than the ancient limit -
+// was not yet comitted, but the process crashed. In this case we expect the
+// chain to detect that it was fast syncing and not delete anything, since we
+// can just pick up directly where we left off.
+func TestLongReorgedFastSyncingRepair(t *testing.T) {
+	testRepair(t, 24, 26, 16, 0, 4, 24, 26)
+}
+
+func testRepair(t *testing.T, canonchain int, sidechain int, freeze uint64, commit int, pivot int, canonretain int, sideretain int) {
```

Comment:
If you use a struct here, instead of individual parameters, the callers can use the field names. I think this would make it a lot clearer.